chore: {:focus} and {:split-here} doc and Verification debugging techniques (#1884)
dafny-lang/dafny: docs/DafnyRef/Attributes.md

> # 22. Attributes

Reviewer: Top-level comment: I appreciate you improving the markdown here and filling out some critical information, but shouldn't we be pushing some of this content into the actual output of `/attrHelp`, both in Boogie and in Dafny?

Reviewer 2: I like that idea (that is, making `/attrHelp` give a URL that points into this manual).

Author: I implemented that idea (just add a permanent link in /attrHelp). I realized that there are a lot of duplicates, so I reimported the documentation about {:extern} from DafnyOptions because it's much better than what I wrote, and I also added the documentation about `{:print}`

Author: Down the road and for a later time, I will probably import others `{:warnShadowing}`, but I really think the documentation should be in one place.

will have its default behavior which is to choose a native type that can hold any
value satisfying the constraints, if possible, otherwise BigInteger is used.
* `{:nativeType true}` - Also gives default ``NewtypeDecl`` behavior,
but gives an error if base type is not integral.
* `{:nativeType false}` - Inhibits using a native type. BigInteger is used.
* `{:nativeType "typename"}` - This form has an native integral
type name as a string literal. Acceptable values are:
   * `"byte"`      8 bits, unsigned
   * `"sbyte"`     8 bits, signed
   * `"ushort"`    16 bits, unsigned
   * `"short"`     16 bits, signed
   * `"uint"`      32 bits, unsigned
   * `"int"`       32 bits, signed
   * `"number"`    53 bits, signed
   * `"ulong"`     64 bits, unsigned
   * `"long"`      64 bits, signed
  If the target compiler
  does not support X, then an error is generated. Also, if, after
  scrutinizing the constraint predicate, the compiler cannot confirm
  that the type's values will fit in X, an error is generated.

### 22.1.3. `{:ignore}` (deprecated)
Ignore the declaration (after checking for duplicate names).

### 22.1.4. `{:extern}` {#sec-extern}

> `{:extern}`  

Author: ```suggestion
```

> `{:extern <s1:string>}`  

Author: ```suggestion
```

> `{:extern <s1:string>, <s2:string>}`  

Author: ```suggestion
```

> 

Author: ```suggestion
```

`{:extern}` is a target-language dependent modifier used:

> * to alter the CompileName of entities such as modules, classes, methods, etc.,

Author: ```suggestion
* to alter the `CompileName` of entities such as modules, classes, methods, etc.,
```

> * to alter the ReferenceName of the entities,

Author: ```suggestion
* to alter the `ReferenceName` of the entities,
```

* to decide how to define external opaque types,
* to decide whether to emit target code or not, and
* to decide whether a declaration is allowed not to have a body.

The `CompileName` is the name for the entity when translating to one of the target languages.

> The `ReferenceName` is the name used to refer to the entity in the target language.

Reviewer: This is confusing to me. My understanding is that the `CompileName` is the unqualified symbol and `ReferenceName` is the qualified symbol.

Author: This is documentation straight from the command line option `/attrHelp`, because of a recent comment of you. I did not make an effort to understand it. I am not sure what "qualified symbol" and "unqualified symbols" mean anyway.
I suggest we do what you suggested, meaning keeping any rewording to this section for later, unless you have specific rewording to suggest that could fit right now and that you are sure about :-)

Reviewer: Ack, missed that this wasn't new content. We'll improve it later.

A common use case of `{:extern}` is to avoid name clashes with existing library functions.

`{:extern}` takes 0, 1, or 2 (possibly empty) string arguments:

> - 0 arguments: Dafny will use the Dafny name as the `CompileName` and not affect the `ReferenceName`

Author: ```suggestion
- `{:extern}`: Dafny will use the Dafny name as the `CompileName` and not affect the `ReferenceName`
```

> - 1 arguments: Dafny will use `s1` as the `CompileName`, and replaces the last portion of the `ReferenceName` by `s1`.

Reviewer: Your `s1` and `s2` symbols don't resolve to anything. :) I would replace "1 arguments" with "`{:extern s1}`" and likewise in the next bullet.

Author: Line 137 and 138 are defining s1 :-) but I agree your approach is better, let me do the changes

> - 1 arguments: Dafny will use `s1` as the `CompileName`, and replaces the last portion of the `ReferenceName` by `s1`.

Author: ```suggestion
- - `{:extern s1}`: Dafny will use `s1` as the `CompileName`, and replaces the last portion of the `ReferenceName` by `s1`.
```

Reviewer: I think you've got an extra `-` here

> - 1 arguments: Dafny will use `s1` as the `CompileName`, and replaces the last portion of the `ReferenceName` by `s1`.

Reviewer: ```suggestion
- `{:extern s1}`: Dafny will use `s1` as the `CompileName`, and replaces the last portion of the `ReferenceName` by `s1`.
```

     When used on an opaque type, s1 is used as a hint as to how to declare that type when compiling.

> - 2 arguments: Dafny will use `s2` as the `CompileName`.

Author: ```suggestion
- `{:extern s1, s2}` Dafny will use `s2` as the `CompileName`.
```

   if (F != null) { Repr := Repr + {F} + F.Repr; }
```

### 22.1.2. `{:nativeType}` {#sec-nativetype}
The `{:nativeType}` attribute may only be used on a ``NewtypeDecl``
where the base type is an integral type. It can take one of the following
forms:

* `{:nativeType}` - With no parameters it has no effect and the ``NewtypeDecl``
have its default behavior which is to choose a native type that can hold any
value satisfying the constraints, if possible, otherwise BigInteger is used.

> * `{:nativeType true}` - Also gives default ``NewtypeDecl`` behavior,

Reviewer 2: To be consistent with other attributes, `{:nativeType}` and `{:nativeType true}` should do the same thing. However, that probably lies outside this PR. For now, it's good that we at least document what the current behavior is.

Author: Issue filed.

but gives an error if base type is not integral.
* `{:nativeType false}` - Inhibits using a native type. BigInteger is used

> for integral types and BitRational for real types.

Reviewer 2: I suggest leaving out "and BitRational for real types", since the text above says `:nativeType` applies only to integral types. (But if these words are kept, then "BitRational" should be corrected to "BigRational".)

* `{:nativeType "typename"}` - This form has an native integral
type name as a string literal. Acceptable values are: "byte",

> "sbyte", "ushort", "short", "uint", "int", "ulong" and "long".

Reviewer 2: I thought that `"number"` is also an acceptable value. (It stands for a 53-bit integer, like in JavaScript.)

Author: I added it this morning.

An error is reported if the given datatype cannot hold all the
values that satisfy the constraint.

### 22.1.3. `{:ignore}`

> Ignore the declaration (after checking for duplicate names).

Reviewer 2: Do the verifier and compilers respect this attribute? It seems like a really strange attribute to me. I don't know (or remember) where it came from, but I would recommend eliminating support for this attribute (in a different PR would be fine).

Author: Boogie does verify this attribute, and it looks like extern trigger some `{:ignore}`. I'll mark it as deprecated

It means that the post-condition may be assumed to be true
without proof. In that case also the body of the function or
method may be omitted.

The `{:axiom}` attribute only prevents Dafny from verifying that the body matches the post-condition.
Dafny still verifies the well-formedness of pre-conditions, of post-conditions, and of the body if provided.
To prevent Dafny from running all these checks, one would use [`{:verify false}`](#sec-verify), which is not recommended.

### 22.2.3. `{:compile}`
The `{:compile}` attribute takes a boolean argument. It may be applied to
any top-level declaration. If that argument is false, then that declaration
will not be compiled at all.
The difference with [`{:extern}`](#sec-extern) is that [`{:extern}`](#sec-extern)
will still emit declaration code if necessary,
whereas `{:compile false}` will just ignore the declaration for compilation purposes.

### 22.2.4. `{:extern <name>}` {#sec-extern-method}
See [`{:extern <name>}`](#sec-extern).

### 22.2.5. `{:fuel X}` {#sec-fuel}
The fuel attributes is used to specify how much "fuel" a function should have,
i.e., how many times the verifier is permitted to unfold its definition.  The

> new `{:fuel}` annotation can be added to the function itself, it which

Reviewer 2: ```suggestion
`{:fuel}` annotation can be added to the function itself, it which
```

   if (F != null) { Repr := Repr + {F} + F.Repr; }
```

### 22.1.4. `{:axiom}` {#sec-axiom}
The `{:axiom}` attribute may be placed on a function or method.
It means that the post-condition may be assumed to be true
without proof. In that case also the body of the function or
method may be omitted.

The `{:axiom}` attribute only prevents Dafny from verifying that the body matches the post-condition.
Dafny still verifies the well-formedness of pre-conditions, of post-conditions, and of the body if provided.
To prevent Dafny from running all these checks, one would use [`{:verify false}`](#sec-verify), which is not recommended.

### 22.1.5. `{:compile}`
The `{:compile}` attribute takes a boolean argument. It may be applied to
any top-level declaration. If that argument is false, then that declaration
will not be compiled at all.
The difference with [`{:extern}`](#sec-extern) is that [`{:extern}`](#sec-extern)
will still emit declaration code if necessary,
whereas `{:compile false}` will just ignore the declaration for compilation purposes.

### 22.1.6. `{:fuel X}` {#sec-fuel}
The fuel attributes is used to specify how much "fuel" a function should have,

> i.e., how many times the SMT solver is permitted to unfold its definition.  The

Reviewer 2: I prefer "verifier" to "SMT solver".

which may depend on the heap, is also valid for any good heap that is either the
same as the current heap, or that is derived from it by heap update operations.

> ### 22.1.9. imported

Reviewer: Does this still exist?

Author: Nope, I checked with @RustanLeino  and it does not exists anymore. People can write their own attributes if they want, but "imported" is not an attribute supported by Dafny

values that satisfy the constraint.

### 22.1.13. `{:opaque}` {#sec-opaque}

> Ordinarily the body of a function is transparent to its users but

Reviewer 3: ```suggestion
Ordinarily, the body of a function is transparent to its users, but
```

}
```

### 22.3.2. `{:split_here}` {#sec-split_here}
`assert {:split_here} X;` splits verification into two [assertion batches](#sec-assertion-batches).
It verifies the code leading to this point (excluded) in a first assertion batch,
and the code leading from this point (included) to the next `{:split_here}` or until the end in a second assertion batch.
It might help with timeouts.
It might also occasionally double-report errors.

```
method doSplitHere(x: bool) returns (y: int) {
  y := 1;
  assert y == 1;   // Assertion in batch #1, Assumption in batch #2
  assert {:split_here} true;
  assert y == 1;   // Assumption in batch #1, Assertion in batch #2
  assert y >= 1;   // Assumption in batch #1, Assertion in batch #2
}
```

### 22.3.3. `{:subsumption n}`
Overrides the `/subsumption` command-line setting for this assertion.

<!--
This is not working in Dafny and shouldn't be documented until so.
### 22.3.4. `{:msg <string>}`
     
Prints <string> rather than the standard message for assertion failure.
Also applicable to requires and ensures declarations.

```
method doFocus(x: bool) returns (y: int) {
  y := 1;
  assert {:msg "y is indivisible"} y != 1;    //
}
```
-->

> 

Reviewer 2: Good to leave this out. I suggest just deleting these lines altogether.

## 22.4. Attributes on variable declarations

### 22.4.1. `{:assumption}`
This attribute can only be placed on a local ghost bool
variable of a method. Its declaration cannot have a rhs, but it is
allowed to participate as the lhs of exactly one assignment of the
form: `b := b && expr;`. Such a variable declaration translates in the
Boogie output to a declaration followed by an `assume b` command.
See [@LeinoWuestholz2015], Section 3, for example uses of the `{:assumption}`
attribute in Boogie.

### 22.4.2. `{:existential}`
Marks a global Boolean variable as existentially quantified. If
used in combination with option `/contractInfer`, Boogie will check
whether there exists a Boolean assignment to the existentials
that makes all verification conditions valid.  Without option

> `/contractInfer` the attribute is ignored.

Reviewer 2: Does this option work with Dafny?

Author: I have no idea, as it's the only documentation that exists about it. It should work however, since attributes are passed onto Boogie. I suggest we remove deprecate this attribute though?

## 22.5. Attributes on quantifier expressions (forall, exists)

### 22.5.1. `{:heapQuantifier}`
The `{:heapQuantifier}` attribute may be used on a [`QuantifierExpression`](#sec-quantifier-expression).
When it appears in a quantifier expression, it is as if a new heap-valued
quantifier variable was added to the quantification. Consider this code
that is one of the invariants of a while loop.

```dafny
invariant forall u {:heapQuantifier} :: f(u) == u + r
```

The quantifier is translated into the following Boogie:

```
(forall q$heap#8: Heap, u#5: int ::
    {:heapQuantifier}
    $IsGoodHeap(q$heap#8) && ($Heap == q$heap#8 || $HeapSucc($Heap, q$heap#8))
       ==> $Unbox(Apply1(TInt, TInt, f#0, q$heap#8, $Box(u#5))): int == u#5 + r#0);
```

What this is saying is that the quantified expression, `f(u) == u + r`,
which may depend on the heap, is also valid for any good heap that is either the
same as the current heap, or that is derived from it by heap update operations.

### 22.5.2. `{:induction}` {#sec-induction-quantifier}
See [`{:induction}`](#sec-induction).

> ### 22.5.3. `{:layerQuantifier}`

Reviewer 2: The word "layer" is old. We refer to this as "fuel" now. I suggest changing the text to use the word "fuel" instead of "layer", perhaps with a comment like `the word "layer" really refers to "fuel"`.

Author: I reopened the issue
https://github.com/dafny-lang/dafny/issues/35

When Dafny is translating a quantified expression, if it has
a `{:layerQuantifier}` attribute an additional quantifier
variable is added to the quantifier bound variables.
This variable as the predefined _LayerType_.
A `{:layerQuantifier}` attribute may be placed on a quantifier expression.
Translation of Dafny into Boogie defines a _LayerType_ which has defined zero and
successor constructors.

The Dafny source has the comment that "if a function is recursive,
then make the reveal lemma quantifier a layerQuantifier."
And in that case it adds the attribute to the quantifier.

> There is no explicit user of the `{:layerQuantifier}` attribute

Reviewer 4: ```suggestion
There is no explicit use of the `{:layerQuantifier}` attribute
```

in the Dafny tests. So I believe this attribute is only used
internally by Dafny and not externally.

TODO: Need more complete explanation of this attribute.
Dafny issue [35](https://github.com/Microsoft/dafny/issues/35) tracks
further effort for this attribute.

### 22.5.4. `{:trigger}` {#sec-trigger}
Trigger attributes are used on quantifiers and comprehensions.
They are used for the verifier to know models of expressions that enable to instantiate the quantifier,

> so that it is not instantiated with anything.

Reviewer 4: This sentence is a little unclear.

Author: I replaced this sentence with a detailed example.

refinement is described.
-->

### 22.1.12. `{:tailrecursion true/false}`

> This attribute is used on method declarations. It has a boolean argument.

Reviewer 2: It can also be used with no arguments, in which the case the argument defaults to `true` (like it does for most/all other attributes that take a boolean argument).

* If `{:tailrecursion true}` was specified but the code does not allow it,
an error message is given.

### 22.1.13. `{:timeLimitMultiplier X}`
This attribute may be placed on a method or function declaration
and has an integer argument. If `{:timeLimitMultiplier X}` was
specified a `{:timelimit Y}` attributed is passed on to Boogie
where `Y` is `X` times either the default verification time limit
for a function or method, or times the value specified by the
Boogie `timelimit` command-line option.

### 22.1.14. `{:trigger}` {#sec-trigger}
Trigger attributes are used on quantifiers and comprehensions.
They are translated into Boogie triggers.

### 22.1.15. `{:typeQuantifier}`
The `{:typeQuantifier}` attribute must be used on a quantifier if it

> quantifies over types.

Reviewer 2: I see the Dafny implementation looks for this attribute, but I doubt it does anything sensible with it. At best, this attribute is deprecated. I suggest changing the text here to say that it is deprecated (and removing the code for it in a separate PR).

## 22.2. Boogie-specific verification attributes
Dafny passes verification attributes that have been specified to its dependency Boogie.
Use the `/attrHelp` option to get the list of attributes of both Dafny and Boogie
recognized by Boogie and their meaning.

### 22.2.1. Verification attributes on top-level declarations

#### 22.2.1.1. `{:ignore}`
Ignore the declaration (after checking for duplicate names).

> #### 22.2.1.2. `{:extern <name>}` {#sec-extern}

Reviewer 2: This section is misplaced. This attribute is used by the Dafny compiler, not by Boogie.

If a ``ClassDecl``, a ``MethodDecl`` or a ``FunctionDecl`` has an `{:extern}` attribute,
then:
* The compiler generally emits nothing for it and will use <name> instead of the dafny provided name in the generated code referencing it.
* `{:extern}` Methods and functions can have [`requires` clauses](#sec-requires-clause) and [`ensures` clauses](#sec-ensures-clause).
* `{:extern}` methods typically do not have any body, since they are unchecked and opaque.
* `{:extern}` Functions can have a body which is used only for verification purposes.

The difference with [`{:axiom}`](#sec-axiom) is that the compiler will still emit code for an [`{:axiom}`](#sec-axiom), if it is a [`function method`, a `method` or a `function by method`](#sec-function-declarations) with a body.

#### 22.2.1.3. `{:checksum <string>}`

> Attach a checksum to be used for verification result caching.

Reviewer 2: This attribute should not be used by Dafny users. It lets Dafny give some directives to Boogie. I suggest removing it from this list.

### 22.2.2. Verification attributes on functions and methods

#### 22.2.2.1. `{:inline N}`
     
Inline a given procedure (can be also used on implementation).
N should be a non-negative number and represents the inlining depth.
With `/inline:assume`, call is replaced with `assume false` once inlining depth is reached.
With `/inline:assert`, call is replaced with `assert false` once inlining depth is reached.
With `/inline:spec`, call is left as is once inlining depth is reached.
With the above three options, methods with the attribute `{:inline N}` are not verified.

> With `/inline:none` the entire attribute is ignored.

Reviewer 2: Boogie's `:inline` attribute expects a called `procedure P` to having a matching `implementation P`. That's not how Dafny generates Boogie code, so `:inline` has no effect on Dafny programs. I suggest removing this section. Alternatively, replace the contents of this section to say that Boogie's `:inline` attribute has no effect on Dafny programs.

Author: I removed this section altogether. Boogie documentation should stay within Boogie, unless it's useful in Dafny.

Assign a unique ID to an implementation to be used for verification
result caching (default: "<impl. name>:0").

#### 22.2.2.10. `{:timeLimit N}`
Set the time limit for a given implementation.

### 22.2.3. Verification attributes on functions

#### 22.2.3.1. `{:builtin "spec"}`
#### 22.2.3.2. `{:bvbuiltin "spec"}`
Rewrite the function to built-in prover function symbol 'fn'.

#### 22.2.3.3. `{:inline}`
`{:inline}` or `{:inline true}` expands a function according to its definition before going to the prover.

#### 22.2.3.4. `{:never_pattern true}`
Terms starting with this function symbol will never be
automatically selected as patterns. It does not prevent them
from being used inside the triggers, and does not affect explicit
trigger annotations. Internally it works by adding `{:nopats ...}`
annotations to quantifiers.

#### 22.2.3.5. `{:identity}`
If the function annotated with `{:identity}` or `{:identity true}` has 1 argument and the use of it has type `X->X` for
some `X`, then the abstract interpreter will treat the function as an
identity function.  Note, the abstract interpreter trusts the
attribute--it does not try to verify that the function really is an

> identity function on the abstract domain.

Reviewer 2: This Boogie attributes don't make sense in a Dafny program. I suggest these not be mentioned in the Dafny reference manual.

### 22.2.4. Verification attributes on variables

#### 22.2.4.1. `{:existential true}`
Marks a global Boolean variable as existentially quantified. If
used in combination with option `/contractInfer`, Boogie will check
whether there exists a Boolean assignment to the existentials
that makes all verification conditions valid.  Without option
`/contractInfer` the attribute is ignored.

### 22.2.5. Verification attributes on assert statements {#sec-verification-attributes-on-assert-statements}

#### 22.2.5.1. `{:focus}` {#sec-focus}
`assert {:focus} X;` splits verification into two [assertion batches](#sec-assertion-batches).
The first problem considers all assertions that are on the bloc containing the `assert {:focus} X;`.
The second problem considers all remaining assertions[^second-focus-assertion-batch].

[^second-focus-assertion-batch]: More precisely, any assertion on an execution path that does not go through the assertion with `{:focus}` will be counted in the second problem, all the others will be counted in the first problem.

For example:
```
method doFocus(x: bool) returns (y: int) {
  y := 1;
  assert y == 1;    // Assumption in batch #1, Assertion in batch #2

>   if(x) {

Reviewer 2: ```suggestion
  if x {
```

* If `{:tailrecursion true}` was specified but the code does not allow it,
an error message is given.

### 22.1.13. `{:timeLimitMultiplier X}`
This attribute may be placed on a method or function declaration
and has an integer argument. If `{:timeLimitMultiplier X}` was
specified a `{:timelimit Y}` attributed is passed on to Boogie
where `Y` is `X` times either the default verification time limit
for a function or method, or times the value specified by the
Boogie `timelimit` command-line option.

### 22.1.14. `{:trigger}` {#sec-trigger}
Trigger attributes are used on quantifiers and comprehensions.
They are translated into Boogie triggers.

### 22.1.15. `{:typeQuantifier}`
The `{:typeQuantifier}` attribute must be used on a quantifier if it
quantifies over types.

## 22.2. Boogie-specific verification attributes
Dafny passes verification attributes that have been specified to its dependency Boogie.

> Use the `/attrHelp` option to get the list of attributes

Reviewer: Doesn't this include attributes specific to Dafny too though?

Author: ```suggestion
Use the `/attrHelp` option to get the list of attributes of both Dafny and Boogie
```

Author: I ended up removing all Boogie-specific attributes that have no meaning for Dafny, and making all attribute explanations related to Dafny

recognized by Boogie and their meaning.

### 22.2.1. Verification attributes on top-level declarations

#### 22.2.1.1. `{:ignore}`
Ignore the declaration (after checking for duplicate names).

#### 22.2.1.2. `{:extern}` {#sec-extern}

If a ``ClassDecl``, a ``MethodDecl`` or a ``FunctionDecl`` has an `{:extern}` attribute,
then:

> * The compiler is going to emit nothing for it, it assumes it exists in the linked files

Reviewer: I don't think this is true in all cases - at least some of the compilers will allow you to declare a body, and the effect of `{:extern}` is only to avoid mangling the name so that other linked files can reference the compiled symbol reliably.

You might want to leave documenting `{:extern}` better out of scope for this already-large PR as it's a bit of a rathole.

Author: I modified "The compiler _generally_ emits nothing for it"
I still prefer to document `{:extern}` now rather than postponing it to another PR. We can always refine it later. The sooner some correct information is available and some wrong information is removed, the better.

* Methods should not have any body, since they are unchecked and opaque.
* Functions can have a body which is used only for verification purposes.

The difference with [`{:axiom}`](#sec-axiom) is that the compiler will still emit code for an [`{:axiom}`](#sec-axiom), if it is a [`function method`, a `method` or a `function by method`](#sec-function-declarations) with a body.

#### 22.2.1.3. `{:checksum <string>}`

Attach a checksum to be used for verification result caching.

### 22.2.2. Verification attributes on functions and methods

#### 22.2.2.1. `{:inline N}`
     
Inline a given procedure (can be also used on implementation).
N should be a non-negative number and represents the inlining depth.
With `/inline:assume`, call is replaced with `assume false` once inlining depth is reached.
With `/inline:assert`, call is replaced with `assert false` once inlining depth is reached.
With `/inline:spec`, call is left as is once inlining depth is reached.
With the above three options, methods with the attribute `{:inline N}` are not verified.
With `/inline:none` the entire attribute is ignored.

#### 22.2.2.2. `{:verify false}` {#sec-verify}
     

> Skip verification of an implementation altogether

Reviewer: ```suggestion
Skip verification of an implementation altogether.
```

Also what does "implementation" mean here? Won't it also skip the well-formedness proof obligations?

Author: Removed reference to Boogie, and described in better detail.

from being used inside the triggers, and does not affect explicit
trigger annotations. Internally it works by adding `{:nopats ...}`
annotations to quantifiers.

#### 22.2.3.5. `{:identity}`
If the function annotated with `{:identity}` or `{:identity true}` has 1 argument and the use of it has type X->X for
some X, then the abstract interpreter will treat the function as an
identity function.  Note, the abstract interpreter trusts the
attribute--it does not try to verify that the function really is an
identity function on the abstract domain.

### 22.2.4. Verification attributes on variables

#### 22.2.4.1. `{:existential true}`
Marks a global Boolean variable as existentially quantified. If
used in combination with option `/contractInfer` Boogie will check
whether there exists a Boolean assignment to the existentials
that makes all verification conditions valid.  Without option
`/contractInfer` the attribute is ignored.

### 22.2.5. Verification attributes on assert statements {#sec-verification-attributes-on-assert-statements}

To understand how to tweak verification,
it is first useful to understand how Dafny verifies functions and methods.

> For every method, Dafny extracts _assertions_, as follows:

Reviewer: And functions, and constructors, and... :)

I'd also work "for example" into this heading in some way, to make sure we're not implying this is exhaustive.

Author: Besides `for example`, I added also assertions for subset types and decreases clause. We have a rich ecosystem of assertions !

* any explicit [`assert` statement](#sec-assert-statement) is _an assertion_.

> * A consecutive pair of lines in a [`calc` statement](#sec-calc-statement) forms _an assertion of their equality_.

Reviewer: Not necessarily equality, you can specify a different chaining operator such as `<=`.

* Every function or method call with a [`requires` clause](#sec-requires-clause) yields _one assertion per clause_
  (special cases such as sequence indexing come with a special require clause that the index is within bounds).
* Assignments `o.f := E;` yield an _assertion_ that `o.f` is allowed by the enclosing [`modifies` clause](#sec-loop-framing).

> * Assignments by predicate `x :| P(x)` yield an _assertion_ that `exists x :: P(x)`.

Reviewer: The reference manual calls this the "assign-such-that operator"

> * Every [`ensures` clause](#sec-ensures-clause) yields an _assertion_ at the end of the method.

Reviewer: Not just the end of the method, on every `return` as well.

* ...

> It is useful to mentally visualize all these assertions as a list (an _assertion batch_) that roughly follows the order in the code[^complexity-path-encoding],

Reviewer: I strongly feel this kind of detailed content should be in the reference manual properly instead, rather than something dumped to the console for a user to read. I'm going to pause reviewing the rest of this to figure that out first.

an error message is given.

### 22.1.15. `{:timeLimitMultiplier X}`
This attribute may be placed on a method or function declaration
and has an integer argument. If `{:timeLimitMultiplier X}` was
specified a `{:timelimit Y}` attributed is passed on to Boogie
where `Y` is `X` times either the default verification time limit
for a function or method, or times the value specified by the
Boogie `timelimit` command-line option.

### 22.1.16. `{:trigger}` {#sec-trigger}
Trigger attributes are used on quantifiers and comprehensions.
They are translated into Boogie triggers.

### 22.1.17. `{:typeQuantifier}`
The `{:typeQuantifier}` attribute must be used on a quantifier if it
quantifies over types.

## 22.2. Boogie-specific verification attributes
Dafny passes verification attributes that have been specified to its dependency Boogie.
Use the "/attrHelp" option to get the list of attributes
recognized by Boogie and their meaning.

### 22.2.1. Verification attributes on top-level declarations

#### 22.2.1.1. `{:ignore}`
Ignore the declaration (after checking for duplicate names).

#### 22.2.1.2. `{:extern}` {#sec-extern}

> If two top-level declarations introduce the same name (for example, two

Reviewer 3: Should we talk about the meaning outside of verification?

constants with the same name or two procedures with the same name), then
Boogie usually produces an error message.  However, if at least one of
the declarations is declared with :extern, one of the declarations is
ignored.  If both declarations are :extern, Boogie arbitrarily chooses
one of them to keep; otherwise, Boogie ignore the :extern declaration
and keeps the other.

#### 22.2.1.3. `{:checksum <string>}`

Attach a checksum to be used for verification result caching.

### 22.2.2. Verification attributes on functions and methods

#### 22.2.2.1. `{:inline N}`
     

> Inline given procedure (can be also used on implementation).

Reviewer 3: ```suggestion
Inline a given procedure (can be also used on implementation).
```

N should be a non-negative number and represents the inlining depth.
With `/inline:assume`, call is replaced with `assume false` once inlining depth is reached.
With `/inline:assert`, call is replaced with `assert false` once inlining depth is reached.
With `/inline:spec`, call is left as is once inlining depth is reached.
With the above three options, methods with the attribute `{:inline N}` are not verified.
With `/inline:none` the entire attribute is ignored.

#### 22.2.2.2. `{:verify false}` {#sec-verify}
     

> Skip verification of an implementation alltogether

Reviewer 3: ```suggestion
Skip verification of an implementation altogether.
```

#### 22.2.2.3. `{:vcs_max_cost N}`
#### 22.2.2.4. `{:vcs_max_splits N}`
#### 22.2.2.5. `{:vcs_max_keep_going_splits N}`
Per-implementation versions of
`/vcsMaxCost`, `/vcsMaxSplits` and `/vcsMaxKeepGoingSplits`.

#### 22.2.2.6. `{:selective_checking true}`
Turn all asserts into assumes except for the ones reachable from
assumptions marked with the attribute `{:start_checking_here}`.
Thus, `assume {:start_checking_here} something;` becomes an inverse
of `assume false;`: the first one disables all verification before
it, and the second one disables all verification after.

#### 22.2.2.7. `{:priority N}`
Assign a positive priority 'N' to an implementation to control the order
in which implementations are verified (default: N = 1).

#### 22.2.2.8. `{:id <string>}`
Assign a unique ID to an implementation to be used for verification
result caching (default: "<impl. name>:0").

#### 22.2.2.9. `{:timeLimit N}`
Set the time limit for a given implementation.

> ### 22.2.3. Verification attributes functions

Author: ```suggestion
### 22.2.3. Verification attributes on functions
```

#### 22.2.3.1. `{:builtin "spec"}`
#### 22.2.3.2. `{:bvbuiltin "spec"}`
Rewrite the function to built-in prover function symbol 'fn'.

#### 22.2.3.3. `{:inline}`

> `{:inline}` or `{:inline true}` expands function according to its definition before going to the prover.

Reviewer 3: ```suggestion
`{:inline}` or `{:inline true}` expands a function according to its definition before going to the prover.
```

#### 22.2.3.4. `{:never_pattern true}`
Terms starting with this function symbol will never be
automatically selected as patterns. It does not prevent them
from being used inside the triggers, and does not affect explicit
trigger annotations. Internally it works by adding `{:nopats ...}`
annotations to quantifiers.

#### 22.2.3.5. `{:identity}`

> If the function annotated with `{:identity}` or `{:identity true}` has 1 argument and the use of it has type X->X for

Reviewer 3: ```suggestion
If the function annotated with `{:identity}` or `{:identity true}` has 1 argument and the use of it has type `X->X` for
```

Author: This was removed because it's Boogie-specific.

> some X, then the abstract interpreter will treat the function as an

Reviewer 3: ```suggestion
some `X`, then the abstract interpreter will treat the function as an
```

identity function.  Note, the abstract interpreter trusts the
attribute--it does not try to verify that the function really is an
identity function.

### 22.2.4. Verification attributes on variables

#### 22.2.4.1. `{:existential true}`
Marks a global Boolean variable as existentially quantified. If

> used in combination with option `/contractInfer` Boogie will check

Reviewer 3: ```suggestion
used in combination with option `/contractInfer`, Boogie will check
```

* Assignments by predicate `x :| P(x)` yield an _assertion_ that `exists x :: P(x)`.
* Every [`ensures` clause](#sec-ensures-clause) yields an _assertion_ at the end of the method.
* ...

It is useful to mentally visualize all these assertions as a list that roughly follows the order in the code[^complexity-path-encoding],
except for `ensures` assertions that appear before in the code but, for verification purposes, would appear at the end.
Thus, to prove or disprove a single assertion within this list, all the assumptions Dafny needs are 1) the global context,
and 2) every preceding assumptions (and previous assertions transformed in assumptions).

[^complexity-path-encoding]: All the complexities of the execution paths (if-then-else, loops, goto, break....) are, down the road and for verification purposes, cleverly encoded with variables recording the paths and guarding assumptions made on each path. In practice, a second clever encoding of variables enables grouping many assertions together, and recovers which assertion is failing based on the value of variables that the SMT solver returns.

> To achieve higher verification performance, Dafny collects all the assertions of one method into one single conjecture (an _assertion batch_) that it sends to the verifier, which tries to prove it correct:

Author: Just saw this series of comments that was automatically hidden by Github when I move this content to another file. Picking them up.

> * If the verifier says it is correct[^smt-encoding], it means that all the assertions hold.

Author: ```suggestion

* If the verifier says it is correct[^smt-encoding], it means that all the assertions hold.
```

* If the verifier returns a counter-example, this counter-example is used to determine both the failing assertion and the failing path.

>   Dafny will then query again the verifier with the assumption that that particular assert is valid, so that it can retrieve other failing assertions happening afterwards.[^caveat-about-assertion-and-assumption]

Reviewer 3: ```suggestion
  In order to retrieve additional failing assertions, Dafny will again query the verifier after turning previously failed assertions into assumptions.[^caveat-about-assertion-and-assumption]
```

* If the verifier returns `unknown` or times out, or even preemptively for difficult assertions or to reduce the chance that the verifier will ‘be confused’ by the many assertions in a large batch, Dafny partitions the assertions up into smaller batches[^smaller-batches]. An extreme case is the use of the `/oneAssertAtATime` command-line option, which causes Dafny to make one batch for each assertion.

[^smt-encoding]: The formula sent to the underlying SMT solver is the negation of the formula that the verifier wants to prove. Hence, if the SMT solver returns "unsat", it means that the SMT formula is always false, meaning the verifier's formula is always true. On the other side, if the SMT solver returns "sat", it means that the SMT formula can be made true with a special variable assignment, which means that the verifier's formula is false under that same variable assignment, meaning it's a counter-example for the verifier.

> [^caveat-about-assertion-and-assumption]: Caveat about assertion and assumption: One big difference between an "assertion transformed in an assumption" and the original "assertion" is that the original "assertion" can unroll functions twice, whereas the "assumed assertion" can unroll them only once. Hence, Dafny can still continue to analyze assertions after a failing assertion without automatically proving "false" (which would make all further assertions to prove automatically).

Reviewer 3: ```suggestion
[^caveat-about-assertion-and-assumption]: Caveat about assertion and assumption: One big difference between an "assertion transformed in an assumption" and the original "assertion" is that the original "assertion" can unroll functions twice, whereas the "assumed assertion" can unroll them only once. Hence, Dafny can still continue to analyze assertions after a failing assertion without automatically proving "false" (which would make all further assertions vacuous).
```

[^smaller-batches]: To create a smaller batch, Dafny duplicates the list of assertions, and ensure every assertion is transformed into an assumption in exactly one list, so that the assertions of the two new lists form a partition of the original list. This results in "easier" formulas for the verifier because it has less to prove, but it takes more overhead because every verification instance have a common set of axioms and there is no knowledge sharing between instances because they run independently.

This is where the following annotations `{:focus}` and `{:split_here}` come it.

> They enables to manually force the verifier to split assertions into two lists, so that both of them can verify usually independently faster than the original problem, although the total verification time, if not using parallel cores, can double at most.

Reviewer 3: Sound we talk about sets instead of lists?

Author: I think it's list is better, because assertions are ordered, and one assertion transforms  into an assumption for all _remaining_ assertions.

#### 22.2.5.1. `{:focus}`
Splits verification into two problems. First problem deletes all paths
that do not have the focus block. Second problem considers the paths
deleted in the first problem and does not contain either the focus block
or any block dominated by it.
For example:

```
method doFocus(x: bool) returns (y: int) {
  y := 1;
  assert y == 1;    // Assert in list #1, Assume in list #2
  if(x) {
    assert {:focus} true;
    y := 2;
    assert y == 2; // Assumed in list #1, Assert in list #2
  } else {
    assert y == 1; // Assert in list #1, Assume in list #2
  }
  assert y >= 1;   // Assert in list #1, Assume in list #2
}
```

#### 22.2.5.2. `{:split_here}`
Verifies code leading to this point and code leading from this point
to the next `{:split_here}` as separate pieces. May help with timeouts.
May also occasionally double-report errors.

```
method doFocus(x: bool) returns (y: int) {
  y := 1;

>   assert y == 1;    // Assert in list #1, Assume in list #2

Reviewer 3: I think it would be helpful to have some deeper nesting here as well.

specified a `{:timelimit Y}` attributed is passed on to Boogie
where `Y` is `X` times either the default verification time limit
for a function or method, or times the value specified by the
Boogie `timelimit` command-line option.

### 22.1.16. `{:trigger}` {#sec-trigger}
Trigger attributes are used on quantifiers and comprehensions.
They are translated into Boogie triggers.

### 22.1.17. `{:typeQuantifier}`
The `{:typeQuantifier}` attribute must be used on a quantifier if it
quantifies over types.

## 22.2. Boogie-specific verification attributes
Dafny passes verification attributes that have been specified to its dependency Boogie.
Use the "/attrHelp" option to get the list of attributes
recognized by Boogie and their meaning.

### 22.2.1. Verification attributes on top-level declarations

#### 22.2.1.1. `{:ignore}`
Ignore the declaration (after checking for duplicate names).

#### 22.2.1.2. `{:extern}` {#sec-extern}
If two top-level declarations introduce the same name (for example, two
constants with the same name or two procedures with the same name), then
Boogie usually produces an error message.  However, if at least one of

> the declarations is declared with :extern, one of the declarations is

Author: Apparently it's arbitrary. I'm asking Rustan about it.

and 2) every preceding assumptions (and previous assertions transformed in assumptions).

[^complexity-path-encoding]: All the complexities of the execution paths (if-then-else, loops, goto, break....) are, down the road and for verification purposes, cleverly encoded with variables recording the paths and guarding assumptions made on each path. In practice, a second clever encoding of variables enables grouping many assertions together, and recovers which assertion is failing based on the value of variables that the SMT solver returns.

To achieve higher verification performance, Dafny collects all the assertions of one method into one single conjecture (an _assertion batch_) that it sends to the verifier, which tries to prove it correct:

* If the verifier says it is correct[^smt-encoding], it means that all the assertions hold.
* If the verifier returns a counter-example, this counter-example is used to determine both the failing assertion and the failing path.
  Dafny will then query again the verifier with the assumption that that particular assert is valid, so that it can retrieve other failing assertions happening afterwards.[^caveat-about-assertion-and-assumption]

> * If the verifier returns `unknown` or times out, or even preemptively for difficult assertions or to reduce the chance that the verifier will ‘be confused’ by the many assertions in a large batch, Dafny partitions the assertions up into smaller batches[^smaller-batches]. An extreme case is the use of the `/oneAssertAtATime` command-line option, which causes Dafny to make one batch for each assertion.

Reviewer 4: Have we renamed this option in the implementation? I like the name you use here better than `/vcsSplitOnEveryAssert`. :)

Author: Ahah indeed we hadn't. I'm adding this special case to Dafny so that it can handle it ! I like it better and easier to remember than `/vcsSplitOnEveryAssert`.

[^smt-encoding]: The formula sent to the underlying SMT solver is the negation of the formula that the verifier wants to prove. Hence, if the SMT solver returns "unsat", it means that the SMT formula is always false, meaning the verifier's formula is always true. On the other side, if the SMT solver returns "sat", it means that the SMT formula can be made true with a special variable assignment, which means that the verifier's formula is false under that same variable assignment, meaning it's a counter-example for the verifier.

[^caveat-about-assertion-and-assumption]: Caveat about assertion and assumption: One big difference between an "assertion transformed in an assumption" and the original "assertion" is that the original "assertion" can unroll functions twice, whereas the "assumed assertion" can unroll them only once. Hence, Dafny can still continue to analyze assertions after a failing assertion without automatically proving "false" (which would make all further assertions to prove automatically).

> [^smaller-batches]: To create a smaller batch, Dafny duplicates the list of assertions, and ensure every assertion is transformed into an assumption in exactly one list, so that the assertions of the two new lists form a partition of the original list. This results in "easier" formulas for the verifier because it has less to prove, but it takes more overhead because every verification instance have a common set of axioms and there is no knowledge sharing between instances because they run independently.

Reviewer 4: I think this isn't quite right. Each assertion appears as an assertion in exactly one list, and every other assertion in that list is transformed into an assumption.

Author: What about the reformulation `Dafny duplicates the list of assertions, and transforms each assertion into an assumption into exactly one list, so that assertions are verified only in one list.`

Reviewer 4: I think it's the opposite of that. "Dafny duplicates the list of assertions (resulting in N lists if there are N assertions), and leaves each assertion in its original form in exactly one list, transforming the rest of that list into assumptions, so that each assertion is verified only in one list."

Author: I don't think this is how it works. What you describe is exactly the behavior of `/vcsSplitOnEveryAssert`, but splitting is more general than that.

This is where the following annotations `{:focus}` and `{:split_here}` come it.
They enables to manually force the verifier to split assertions into two lists, so that both of them can verify usually independently faster than the original problem, although the total verification time, if not using parallel cores, can double at most.

> 

Reviewer 4: I'm not sure it's true that it can double at most. In some weird cases, I think it's possible for splitting verification into batches to slow things down dramatically. Usually it's not worse than doubling (and often it's much better), but I think it's possible for it to be.

Author: I replaced `can double at most.` with `might increase.`

Reviewer 4: Sounds good!

#### 22.2.5.1. `{:focus}`

> Splits verification into two problems. First problem deletes all paths

Reviewer 4: "First problem" -> "The first problem" and "Second problem" -> "The second problem"

that do not have the focus block. Second problem considers the paths
deleted in the first problem and does not contain either the focus block
or any block dominated by it.
For example:

```
method doFocus(x: bool) returns (y: int) {
  y := 1;
  assert y == 1;    // Assert in list #1, Assume in list #2
  if(x) {
    assert {:focus} true;
    y := 2;

>     assert y == 2; // Assumed in list #1, Assert in list #2

Reviewer 4: You say "assumed" here but "assume" everywhere else.

  } else {
    assert y == 1; // Assert in list #1, Assume in list #2
  }
  assert y >= 1;   // Assert in list #1, Assume in list #2
}
```

#### 22.2.5.2. `{:split_here}`
Verifies code leading to this point and code leading from this point
to the next `{:split_here}` as separate pieces. May help with timeouts.
May also occasionally double-report errors.

```
method doFocus(x: bool) returns (y: int) {
  y := 1;
  assert y == 1;    // Assert in list #1, Assume in list #2
  assert {:split_here} true;

>   assert y == 1;   // Assumed in list #1, Assert in list #2

Reviewer 4: Ditto "assumed" vs. "assume".

Per-method version of the command-line option `/vcsMaxKeepGoingSplits`.
If set to more than 1, activates the _keep going mode_ where, after the first round of splitting,
[assertion batches](#sec-assertion-batches) that timed out are split into <n> [assertion batches](#sec-assertion-batches) and retried
until we succeed proving them, or there is only one
single assertion that it timeouts (in which
case error is reported for that assertion).
Defaults to 1.
If [`{:vcs_split_on_every_assert}`](#sec-vcs_split_on_every_assert) is set, then this parameter is useless.

### 22.2.18. `{:vcs_max_splits N}` {#sec-vcs_max_splits}

Per-method version of the command-line option `/vcsMaxSplits`.
Maximal number of [assertion batches](#sec-assertion-batches) generated for this method.
In [keep-going mode](#sec-vcs_max_keep_going_splits), only applies to the first round.
Defaults to 1.
If [`{:vcs_split_on_every_assert}`](#sec-vcs_split_on_every_assert) is set, then this parameter is useless.

### 22.2.19. `{:vcs_split_on_every_assert}` {#sec-vcs_split_on_every_assert}
Per-method version of the command-line option `/vcsSplitOnEveryAssert`.

> In the first verification round, this option will split the original [assertion batch](#sec-assertion-batches)

Reviewer: Again my understanding is that if you JUST use `{:vcs_split_on_every_assert}`, there is only one round of verification. Let's confirm that.

> In the first verification round, this option will split the original [assertion batch](#sec-assertion-batches)

Author: ```suggestion
In the first and only verification round, this option will split the original [assertion batch](#sec-assertion-batches)
```

into one assertion batch per assertion.
This is mostly helpful for debugging which assertion is taking the most time to prove, e.g. to profile them.

## 22.3. Attributes on assertions, preconditions and postconditions {#sec-verification-attributes-on-assert-statements}

### 22.3.1. `{:focus}` {#sec-focus}
`assert {:focus} X;` splits verification into two [assertion batches](#sec-assertion-batches).
The first batch considers all assertions that are not on the block containing the `assert {:focus} X;`
The second batch considers all assertions that are on the block containing the `assert {:focus} X;` and those that will _always_ follow afterwards.
Hence, it might also occasionally double-report errors.
If you truly want a split on the batches, prefer [`{:split_here}`](#sec-split_here).

> Here are two examples illustrating how `{:focus}` works, where `--` in the comments stands for `Assumption`:

Reviewer: Love the way you've illustrated these concepts!

  if x {
    assert y <= 1;             // Assertion  --          --
    assert {:split_here} true; // --         Assertion   --
    assert y <= 2;             // --         Assertion   --
    assert {:split_here} true; // --         --          Assertion
    if x {
      assert y == 1;           // --         --          Assertion
    } else {
      assert y >= 1;           // --         --          Assertion
    }
  } else {
    assert y <= 3;             // Assertion  --          --
  }
  assert y >= -1;              // Assertion  --          --
}
```

### 22.3.3. `{:subsumption n}`
Overrides the `/subsumption` command-line setting for this assertion.

## 22.4. Attributes on variable declarations

### 22.4.1. `{:assumption}`
This attribute can only be placed on a local ghost bool
variable of a method. Its declaration cannot have a rhs, but it is
allowed to participate as the lhs of exactly one assignment of the
form: `b := b && expr;`. Such a variable declaration translates in the
Boogie output to a declaration followed by an `assume b` command.
See [@LeinoWuestholz2015], Section 3, for example uses of the `{:assumption}`
attribute in Boogie.

### 22.4.2. `{:existential}`

> Marks a global Boolean variable as existentially quantified. If

Reviewer: There is no such thing as a global variable in Dafny, so we should leave this one out.

used in combination with option `/contractInfer`, Boogie will check
whether there exists a Boolean assignment to the existentials
that makes all verification conditions valid.  Without option
`/contractInfer` the attribute is ignored.

## 22.5. Attributes on quantifier expressions (forall, exists)

> ### 22.5.1. `{:heapQuantifier}`

Reviewer: This one is meaningless to me as a Dafny programmer AFAICT, so I'd skip it as well.

Author: I can't skip it because it has some meaning, and some specific translation to Boogie, so even if I don't know how to use it, I have no reason to believe it can't be useful for someone dealing with the heap.

Reviewer: Fair enough, I realize that lifting the content up here rather than just saying "here's the dump of attributes Boogie understands, IDK you figure it out" :) means places where we haven't mapped things back to Dafny are more obvious. I'm satisfied we're making forward progress with this at least.
